subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -150,13 +150,33 @@ protected void compile(InputChanges inputChanges) {
         checkGroovyClasspathIsNonEmpty();
         warnIfCompileAvoidanceEnabled();
 
-        if (inputChanges != null && getOptions().isIncremental()) {
-            doIncrementalCompile(inputChanges);
+        GroovyJavaJointCompileSpec spec = createSpec();
+
+        if (inputChanges != null && enableIncrementalCompilation(spec)) {
```

Comment:
There are still a few calls to `enableIncrementalCompilation()`. We should only have one call, and the method should only use fields from `GroovyCompile`. That way, we can create an incremental spec and a non-incremental spec.

Reply:
`enableIncrementalCompilation()` depends on `spec.getSourceRoots()`, which is not easily calculated, so I'll keep the `enableIncrementalCompilation(GroovyJavaJointCompileSpec spec)`, but add a default method `GroovyJavaJointCompileSpec.incrementalCompilationEnabled()` to `GroovyJavaJointCompileSpec` so everyone can simply call this method.